external/llvm-project/mlir/lib/Dialect/MemRef/IR/MemRefOps.cpp

```diff
@@ -2044,6 +2045,146 @@ void ReinterpretCastOp::getCanonicalizationPatterns(RewritePatternSet &results,
   results.add<ReinterpretCastOpExtractStridedMetadataFolder>(context);
 }
 
+//===----------------------------------------------------------------------===//
+// ReinterpretElementsOp
+//===----------------------------------------------------------------------===//
+
+void ReinterpretElementsOp::getAsmResultNames(
+    function_ref<void(Value, StringRef)> setNameFn) {
+  setNameFn(getResult(), "reinterpret_elements");
+}
+
+/// If the in-memory bitwidth of `type` can be determined - that is, if it is a
+/// scalar or a 1D or 0D vector, return it. Only types that have such a bitwidth
+/// may be input or result element types to element reinterpretation unless
+/// alignment checks are off.
+static std::optional<int64_t> bitwidthForElementRenterpretation(Type type) {
+  if (type.isIntOrFloat())
+    return type.getIntOrFloatBitWidth();
+  if (auto vecType = type.dyn_cast<VectorType>()) {
+    if (!vecType.hasRank() || !vecType.hasStaticShape() ||
+        vecType.getNumScalableDims() > 0 || vecType.getRank() > 1)
+      return std::nullopt;
+    Type vecElem = vecType.getElementType();
+    if (!vecElem.isIntOrFloat())
+      return std::nullopt;
+    return vecType.getNumElements() * vecElem.getIntOrFloatBitWidth();
+  }
+  return std::nullopt;
+}
+
+void ReinterpretElementsOp::build(OpBuilder &odsBuilder, OperationState &state,
+                                  Value source, Type newElementType) {
+  state.addOperands(source);
+  MemRefType sourceType = source.getType().dyn_cast<MemRefType>();
+  if (!sourceType) {
+    // Bail in a form that'll quickly trip the verifier.
+    state.addTypes(newElementType);
+    return;
+  }
+  int64_t sourceWidth =
+      bitwidthForElementRenterpretation(sourceType.getElementType())
+          .value_or(1);
+  int64_t destWidth =
+      bitwidthForElementRenterpretation(newElementType).value_or(1);
+  // Best-effort generation of the new dimension lenght - if things are too
+  // long, or don't divide evenly, the verifier will complain.
+  int64_t scaleFactor = std::max(1L, sourceWidth / destWidth);
+  if (scaleFactor == 1)
+    return state.addTypes(sourceType.cloneWith(std::nullopt, newElementType));
+  SmallVector<int64_t> newShape(sourceType.getShape());
+  newShape.push_back(scaleFactor);
+  auto resultType =
+      MemRefType::get(newShape, newElementType, MemRefLayoutAttrInterface{},
+                      sourceType.getMemorySpace());
+  return state.addTypes(resultType);
```

Comment:
Is this fine? I mean you are returning something here but the function signature is `void(..)`

Reply:
Yeah, I'm returning the result of the function, which is also of type `void`